app/vlinsert/loki/loki_json.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/protoparser/common"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/writeconcurrencylimiter"
+
+	"github.com/VictoriaMetrics/metrics"
+)
+
+var (
+	rowsIngestedTotalJSON = metrics.NewCounter(`vl_rows_ingested_total{type="loki", format="json"}`)
+	parserPool            fastjson.ParserPool
+)
+
+func handleJSON(r *http.Request, w http.ResponseWriter) bool {
+	contentType := r.Header.Get("Content-Type")
+	reader := r.Body
+	if contentType == "gzip" {
+		zr, err := common.GetGzipReader(reader)
+		if err != nil {
+			httpserver.Errorf(w, r, "cannot read gzipped request: %s", err)
+			return true
+		}
+		defer common.PutGzipReader(zr)
+		reader = zr
+	}
+
+	cp, err := getCommonParams(r)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot parse request: %s", err)
+		return true
+	}
+	lr := logstorage.GetLogRows(cp.StreamFields, cp.IgnoreFields)
+
+	processLogMessage := cp.GetProcessLogMessageFunc(lr)
+	n, err := processJSONRequest(reader, processLogMessage)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot decode loki request: %s", err)
+		return true
+	}
+	rowsIngestedTotalJSON.Add(n)
+	return true
+}
+
+func processJSONRequest(r io.Reader, processLogMessage func(timestamp int64, fields []logstorage.Field)) (int, error) {
+	wcr := writeconcurrencylimiter.GetReader(r)
+	defer writeconcurrencylimiter.PutReader(wcr)
+
+	bytes, err := io.ReadAll(wcr)
+	if err != nil {
+		return 0, fmt.Errorf("cannot read request body: %w", err)
+	}
+
+	p := parserPool.Get()
+	defer parserPool.Put(p)
+	v, err := p.ParseBytes(bytes)
+	if err != nil {
+		return 0, fmt.Errorf("cannot parse request body: %w", err)
+	}
+
+	rowsIngested := 0
+	for stIdx, st := range v.GetArray("streams") {
+		// `stream` contains labels for the stream.
+		// Labels are same for all entries in the stream.
+		logFields := st.GetObject("stream")
+		if logFields == nil {
+			logger.Warnf("missing streams field from %q", st)
+			logFields = &fastjson.Object{}
+		}
+		commonFields := make([]logstorage.Field, logFields.Len()+1)
+		i := 0
+		logFields.Visit(func(k []byte, v *fastjson.Value) {
+			sfName := bytesutil.ToUnsafeString(k)
+			sfValue := bytesutil.ToUnsafeString(v.GetStringBytes())
+			commonFields[i] = logstorage.Field{
+				Name:  sfName,
+				Value: sfValue,
+			}
+			i++
+		})
+		msgFieldIdx := logFields.Len()
+		commonFields[msgFieldIdx].Name = msgField
+
+		for idx, v := range st.GetArray("values") {
+			vs := v.GetArray()
+			if len(vs) != 2 {
+				return rowsIngested, fmt.Errorf("unexpected number of values in stream %d line %d: %q; got %d; want %d", stIdx, idx, v, len(vs), 2)
+			}
+
+			tsString := bytesutil.ToUnsafeString(vs[0].GetStringBytes())
+			tsString = strings.Trim(tsString, `"`)
+			ts, err := parseLokiTimestamp(tsString)
+			if err != nil {
+				return rowsIngested, fmt.Errorf("cannot parse timestamp in stream %d line %d: %q: %s", stIdx, idx, vs, err)
```

Comment:
Should we stop ingestion on first error? Data may be already added to the storage. And it leads to the partial inserts.

Reply:
This is implemented to follow logic present in other ingestion protocols, e.g. elasticsearch - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/1cc06e39cd74a916bb166bfaff6bc0bfc80d6a15/app/vlinsert/elasticsearch/elasticsearch.go#L162-L210
Other protocols stop ingestion on first error, so this one also stops it for consistency